src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingListener.java

```diff
@@ -0,0 +1,190 @@
+package com.baulsupp.oksocial.tracing;
+
+import brave.Span;
+import brave.Tracer;
+import brave.http.HttpTracing;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.Proxy;
+import java.util.List;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import okhttp3.Call;
+import okhttp3.Connection;
+import okhttp3.EventListener;
+import okhttp3.Handshake;
+import okhttp3.Protocol;
+import zipkin.TraceKeys;
+
+public class ZipkinTracingListener extends EventListener {
+  private final Call call;
+  private final Tracer tracer;
+  private final HttpTracing tracing;
+
+  private Span connectSpan;
+  private Span dnsSpan;
+  private boolean connectionEvent = false;
+  private Span callSpan;
+  private Tracer.SpanInScope spanInScope;
+  private Span requestSpan;
+  private Span responseSpan;
+  private Span secureConnectSpan;
+
+  public ZipkinTracingListener(Call call, Tracer tracer, HttpTracing tracing) {
+    this.call = call;
+    this.tracer = tracer;
+    this.tracing = tracing;
+  }
+
+  @Override public void fetchStart(Call call) {
+    callSpan = tracer.newTrace().name("http").start();
+
+    if (callSpan.isNoop()) {
+      return;
+    }
+
+    callSpan.tag(TraceKeys.HTTP_PATH, call.request().url().encodedPath());
+    callSpan.tag(TraceKeys.HTTP_METHOD, call.request().method());
+    callSpan.tag(TraceKeys.HTTP_HOST, call.request().url().host());
+    callSpan.kind(Span.Kind.CLIENT);
+
+    spanInScope = tracer.withSpanInScope(callSpan);
+  }
+
+  @Override public void fetchEnd(Call call, Throwable throwable) {
+    if (callSpan.isNoop()) {
+      return;
+    }
+
+    callSpan.finish();
+    spanInScope.close();
+
+    if (throwable != null) {
+      callSpan.annotate(throwable.toString());
```

Comment:
usually we add the tag "error" vs annotate (if it is unrecoverable), and something with the word "error" in it as an annotation if it is recoverable.

tag("error" colors the span red
annotate("...error..." colors the span yellow